File: dsa-problems/Arrays/solutions/python/ARR-013-tool-frequency-top-k-decay.py
import sys
import math

def top_k_with_decay(events: list[tuple[int, int]], now: int, D: int, k: int) -> list[int]:
    scores = {}
    
    for val, t in events:
        term = math.exp(-(now - t) / D)
        scores[val] = scores.get(val, 0.0) + term
        
    # Convert to list of (score, val)
    # Sort key: (-score, val) -> Descending score, Ascending val
    items = []
    for val, score in scores.items():
        items.append((-score, val))
        
    items.sort()
    
    result = []
    for i in range(min(k, len(items))):
        result.append(items[i][1]) # Append val
        
    return result

def main():
    n = int(input())
    events = []
    for _ in range(n):
        value, timestamp = map(int, input().split())
        events.append((value, timestamp))
    now, D, k = map(int, input().split())

    result = top_k_with_decay(events, now, D, k)
    print(" ".join(map(str, result)))

if __name__ == "__main__":
    main()

File: dsa-problems/Arrays/solutions/javascript/ARR-013-tool-frequency-top-k-decay.js
const readline = require("readline");

class Solution {
  topKWithDecay(events, now, D, k) {
    const scores = new Map();
    
    for (const [v, t] of events) {
      const term = Math.exp(-(now - t) / D);
      scores.set(v, (scores.get(v) || 0) + term);
    }
    
    const items = [];
    for (const [v, s] of scores.entries()) {
      items.push({ value: v, score: s });
    }
    
    // Sort: Descending Score, Ascending Value
    items.sort((a, b) => {
      if (Math.abs(b.score - a.score) > 1e-9) {
        return b.score - a.score;
      }
      return a.value - b.value;
    });
    
    const result = [];
    for (let i = 0; i < k && i < items.length; i++) {
      result.push(items[i].value);
    }
    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(line.trim()));
rl.on("close", () => {
    if (data.length === 0) return;
    const tokens = data.join(" ").split(/\s+/);
    if (tokens.length === 0 || tokens[0] === "") return;
    
    let ptr = 0;
    const n = Number(tokens[ptr++]);
    const events = [];
    for (let i = 0; i < n; i++) {
      const v = Number(tokens[ptr++]);
      const t = Number(tokens[ptr++]);
      events.push([v, t]);
    }
    
    const now = Number(tokens[ptr++]);
    const D = Number(tokens[ptr++]);
    const k = Number(tokens[ptr++]);
    
    const solution = new Solution();
    const result = solution.topKWithDecay(events, now, D, k);
    console.log(result.join(" "));
});
